File: swift/Sources/MaterialColorUtilities/Quantize/QuantizerWu.swift
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation

public class QuantizerWu: Quantizer {
  var weights: [Int] = []
  var momentsR: [Int] = []
  var momentsG: [Int] = []
  var momentsB: [Int] = []
  var moments: [Double] = []
  var cubes: [Box] = []

  // A histogram of all the input colors is constructed. It has the shape of a
  // cube. The cube would be too large if it contained all 16 million colors:
  // historical best practice is to use 5 bits  of the 8 in each channel,
  // reducing the histogram to a volume of ~32,000.
  static let indexBits: Int = 5
  static let maxIndex: Int = 32
  static let sideLength: Int = 33
  static let totalSize: Int = 35937

  public func quantize(
    _ pixels: [Int], _ maxColors: Int, returnInputPixelToClusterPixel: Bool = false
  )
    -> QuantizerResult
  {
    let result = QuantizerMap().quantize(pixels, maxColors)

    var colorToPopulation = result.colorToCount.map({ (color: $0.key, population: $0.value) })
      .sorted(by: { $0.population > $1.population })

    constructHistogram(colorToPopulation)
    computeMoments()
    let createBoxesResult = createBoxes(maxColors)
    let results = createResult(createBoxesResult.resultCount)
    var colorToCount: [Int: Int] = [:]
    for e in results {
      colorToCount[e] = 0
    }
    return QuantizerResult(colorToCount)
  }

  public static func getIndex(_ r: Int, _ g: Int, _ b: Int) -> Int {
    let indexBits = QuantizerWu.indexBits
    let red = r << (indexBits * 2)
    let green = r << (indexBits + 1)
    let blue = g << indexBits
    return red + green + blue + r + g + b
  }

  public func constructHistogram(_ pixels: [(color: Int, population: Int)]) {
    weights = [Int](repeating: 0, count: QuantizerWu.totalSize)
    momentsR = [Int](repeating: 0, count: QuantizerWu.totalSize)
    momentsG = [Int](repeating: 0, count: QuantizerWu.totalSize)
    momentsB = [Int](repeating: 0, count: QuantizerWu.totalSize)
    moments = [Double](repeating: 0, count: QuantizerWu.totalSize)
    for entry in pixels {
      let pixel = entry.color
      let count = entry.population

      let red = ColorUtils.redFromArgb(pixel)
      let green = ColorUtils.greenFromArgb(pixel)
      let blue = ColorUtils.blueFromArgb(pixel)

      let bitsToRemove = 8 - QuantizerWu.indexBits
      let iR = (red >> bitsToRemove) + 1
      let iG = (green >> bitsToRemove) + 1
      let iB = (blue >> bitsToRemove) + 1
      let index = QuantizerWu.getIndex(iR, iG, iB)

      weights[index] += count
      momentsR[index] += (red * count)
      momentsG[index] += (green * count)
      momentsB[index] += (blue * count)
      moments[index] +=
        (Double(count) * (Double(red * red) + Double(green * green) + Double(blue * blue)))
    }
  }

  public func computeMoments() {
    for r in 1..<QuantizerWu.sideLength {
      var area = [Int](repeating: 0, count: QuantizerWu.sideLength)
      var areaR = [Int](repeating: 0, count: QuantizerWu.sideLength)
      var areaG = [Int](repeating: 0, count: QuantizerWu.sideLength)
      var areaB = [Int](repeating: 0, count: QuantizerWu.sideLength)
      var area2 = [Double](repeating: 0, count: QuantizerWu.sideLength)
      for g in 1..<QuantizerWu.sideLength {
        var line: Int = 0
        var lineR: Int = 0
        var lineG: Int = 0
        var lineB: Int = 0
        var line2: Double = 0
        for b in 1..<QuantizerWu.sideLength {
          let index = QuantizerWu.getIndex(r, g, b)
          line += weights[index]
          lineR += momentsR[index]
          lineG += momentsG[index]
          lineB += momentsB[index]
          line2 += moments[index]

          area[b] += line
          areaR[b] += lineR
          areaG[b] += lineG
          areaB[b] += lineB
          area2[b] += line2

          let previousIndex = QuantizerWu.getIndex(r - 1, g, b)
          weights[index] = weights[previousIndex] + area[b]
          momentsR[index] = momentsR[previousIndex] + areaR[b]
          momentsG[index] = momentsG[previousIndex] + areaG[b]
          momentsB[index] = momentsB[previousIndex] + areaB[b]
          moments[index] = moments[previousIndex] + area2[b]
        }
      }
    }
  }

  public func createBoxes(_ maxColorCount: Int) -> CreateBoxesResult {
    cubes = []
    for _ in 0..<maxColorCount {
      cubes.append(Box())
    }
    cubes[0] = Box(
      r0: 0,
      r1: QuantizerWu.maxIndex,
      g0: 0,
      g1: QuantizerWu.maxIndex,
      b0: 0,
      b1: QuantizerWu.maxIndex,
      vol: 0
    )

    var volumeVariance = [Double](repeating: 0, count: maxColorCount)
    var next = 0
    var generatedColorCount = maxColorCount
    var i = 1
    while i < maxColorCount {
      if cut(cubes[next], cubes[i]) {
        volumeVariance[next] = (cubes[next].vol > 1) ? variance(cubes[next]) : 0
        volumeVariance[i] = (cubes[i].vol > 1) ? variance(cubes[i]) : 0
      } else {
        volumeVariance[next] = 0
        i -= 1
      }

      next = 0
      var temp = volumeVariance[0]
      var j = 1
      while j <= i {
        if volumeVariance[j] > temp {
          temp = volumeVariance[j]
          next = j
        }
        j += 1
      }
      if temp <= 0 {
        generatedColorCount = i + 1
        break
      }
      i += 1
    }

    return CreateBoxesResult(
      requestedCount: maxColorCount,
      resultCount: generatedColorCount
    )
  }

  public func createResult(_ colorCount: Int) -> [Int] {
    var colors: [Int] = []
    for i in 0..<colorCount {
      let cube = cubes[i]
      let weight = QuantizerWu.volume(cube, weights)
      if weight > 0 {
        let r = Int(round(Double(QuantizerWu.volume(cube, momentsR)) / Double(weight)))
        let g = Int(round(Double(QuantizerWu.volume(cube, momentsG)) / Double(weight)))
        let b = Int(round(Double(QuantizerWu.volume(cube, momentsB)) / Double(weight)))
        let color = ColorUtils.argbFromRgb(r, g, b)
        colors.append(color)
      }
    }
    return colors
  }

  public func variance(_ cube: Box) -> Double {
    let dr = QuantizerWu.volume(cube, momentsR)
    let dg = QuantizerWu.volume(cube, momentsG)
    let db = QuantizerWu.volume(cube, momentsB)
    let xx =
      moments[QuantizerWu.getIndex(cube.r1, cube.g1, cube.b1)]
      - moments[QuantizerWu.getIndex(cube.r1, cube.g1, cube.b0)]
      - moments[QuantizerWu.getIndex(cube.r1, cube.g0, cube.b1)]
      + moments[QuantizerWu.getIndex(cube.r1, cube.g0, cube.b0)]
      - moments[QuantizerWu.getIndex(cube.r0, cube.g1, cube.b1)]
      + moments[QuantizerWu.getIndex(cube.r0, cube.g1, cube.b0)]
      + moments[QuantizerWu.getIndex(cube.r0, cube.g0, cube.b1)]
      - moments[QuantizerWu.getIndex(cube.r0, cube.g0, cube.b0)]

    let hypotenuse = (dr * dr + dg * dg + db * db)
    let volume_ = QuantizerWu.volume(cube, weights)
    return xx - Double(hypotenuse / volume_)
  }

  public func cut(_ one: Box, _ two: Box) -> Bool {
    let wholeR = QuantizerWu.volume(one, momentsR)
    let wholeG = QuantizerWu.volume(one, momentsG)
    let wholeB = QuantizerWu.volume(one, momentsB)
    let wholeW = QuantizerWu.volume(one, weights)

    let maxRResult = maximize(
      one, Direction.red, one.r0 + 1, one.r1, wholeR, wholeG, wholeB, wholeW)
    let maxGResult = maximize(
      one, Direction.green, one.g0 + 1, one.g1, wholeR, wholeG, wholeB, wholeW)
    let maxBResult = maximize(
      one, Direction.blue, one.b0 + 1, one.b1, wholeR, wholeG, wholeB, wholeW)

    var cutDirection: Direction
    let maxR = maxRResult.maximum
    let maxG = maxGResult.maximum
    let maxB = maxBResult.maximum
    if maxR >= maxG && maxR >= maxB {
      cutDirection = Direction.red
      if maxRResult.cutLocation < 0 {
        return false
      }
    } else if maxG >= maxR && maxG >= maxB {
      cutDirection = Direction.green
    } else {
      cutDirection = Direction.blue
    }

    two.r1 = one.r1
    two.g1 = one.g1
    two.b1 = one.b1

    switch cutDirection {
    case Direction.red:
      one.r1 = maxRResult.cutLocation
      two.r0 = one.r1
      two.g0 = one.g0
      two.b0 = one.b0
      break
    case Direction.green:
      one.g1 = maxGResult.cutLocation
      two.r0 = one.r0
      two.g0 = one.g1
      two.b0 = one.b0
      break
    case Direction.blue:
      one.b1 = maxBResult.cutLocation
      two.r0 = one.r0
      two.g0 = one.g0
      two.b0 = one.b1
      break
    }

    one.vol = (one.r1 - one.r0) * (one.g1 - one.g0) * (one.b1 - one.b0)
    two.vol = (two.r1 - two.r0) * (two.g1 - two.g0) * (two.b1 - two.b0)
    return true
  }

  public func maximize(
    _ cube: Box, _ direction: Direction, _ first: Int, _ last: Int, _ wholeR: Int, _ wholeG: Int,
    _ wholeB: Int, _ wholeW: Int
  ) -> MaximizeResult {
    let bottomR = QuantizerWu.bottom(cube, direction, momentsR)
    let bottomG = QuantizerWu.bottom(cube, direction, momentsG)
    let bottomB = QuantizerWu.bottom(cube, direction, momentsB)
    let bottomW = QuantizerWu.bottom(cube, direction, weights)

    var max: Double = 0
    var cut: Int = -1

    for i in first..<last {
      var halfR = bottomR + QuantizerWu.top(cube, direction, i, momentsR)
      var halfG = bottomG + QuantizerWu.top(cube, direction, i, momentsG)
      var halfB = bottomB + QuantizerWu.top(cube, direction, i, momentsB)
      var halfW = bottomW + QuantizerWu.top(cube, direction, i, weights)

      if halfW == 0 {
        continue
      }

      var tempNumerator = Double((halfR * halfR) + (halfG * halfG) + (halfB * halfB))
      var tempDenominator = Double(halfW)
      var temp = tempNumerator / tempDenominator

      halfR = wholeR - halfR
      halfG = wholeG - halfG
      halfB = wholeB - halfB
      halfW = wholeW - halfW

      if halfW == 0 {
        continue
      }

      tempNumerator = Double((halfR * halfR) + (halfG * halfG) + (halfB * halfB))
      tempDenominator = Double(halfW)
      temp += (tempNumerator / tempDenominator)

      if temp > max {
        max = temp
        cut = i
      }
    }
    return MaximizeResult(cutLocation: cut, maximum: max)
  }

  public static func volume(_ cube: Box, _ moment: [Int]) -> Int {
    return
      (moment[getIndex(cube.r1, cube.g1, cube.b1)] - moment[getIndex(cube.r1, cube.g1, cube.b0)]
      - moment[getIndex(cube.r1, cube.g0, cube.b1)] + moment[getIndex(cube.r1, cube.g0, cube.b0)]
      - moment[getIndex(cube.r0, cube.g1, cube.b1)] + moment[getIndex(cube.r0, cube.g1, cube.b0)]
      + moment[getIndex(cube.r0, cube.g0, cube.b1)] - moment[getIndex(cube.r0, cube.g0, cube.b0)])
  }

  public static func bottom(_ cube: Box, _ direction: Direction, _ moment: [Int]) -> Int {
    switch direction {
    case Direction.red:
      return -moment[getIndex(cube.r0, cube.g1, cube.b1)]
        + moment[getIndex(cube.r0, cube.g1, cube.b0)] + moment[getIndex(cube.r0, cube.g0, cube.b1)]
        - moment[getIndex(cube.r0, cube.g0, cube.b0)]
    case Direction.green:
      return -moment[getIndex(cube.r1, cube.g0, cube.b1)]
        + moment[getIndex(cube.r1, cube.g0, cube.b0)] + moment[getIndex(cube.r0, cube.g0, cube.b1)]
        - moment[getIndex(cube.r0, cube.g0, cube.b0)]
    case Direction.blue:
      return -moment[getIndex(cube.r1, cube.g1, cube.b0)]
        + moment[getIndex(cube.r1, cube.g0, cube.b0)] + moment[getIndex(cube.r0, cube.g1, cube.b0)]
        - moment[getIndex(cube.r0, cube.g0, cube.b0)]
    }
  }

  public static func top(_ cube: Box, _ direction: Direction, _ position: Int, _ moment: [Int])
    -> Int
  {
    switch direction {
    case Direction.red:
      return
        (moment[getIndex(position, cube.g1, cube.b1)] - moment[getIndex(position, cube.g1, cube.b0)]
        - moment[getIndex(position, cube.g0, cube.b1)]
        + moment[getIndex(position, cube.g0, cube.b0)])
    case Direction.green:
      return
        (moment[getIndex(cube.r1, position, cube.b1)] - moment[getIndex(cube.r1, position, cube.b0)]
        - moment[getIndex(cube.r0, position, cube.b1)]
        + moment[getIndex(cube.r0, position, cube.b0)])
    case Direction.blue:
      return
        (moment[getIndex(cube.r1, cube.g1, position)] - moment[getIndex(cube.r1, cube.g0, position)]
        - moment[getIndex(cube.r0, cube.g1, position)]
        + moment[getIndex(cube.r0, cube.g0, position)])
    }
  }
}

public enum Direction {
  case red
  case green
  case blue
}

public class MaximizeResult {
  // < 0 if cut impossible
  var cutLocation: Int
  var maximum: Double

  init(cutLocation: Int, maximum: Double) {
    self.cutLocation = cutLocation
    self.maximum = maximum
  }
}

public class CreateBoxesResult {
  var requestedCount: Int
  var resultCount: Int

  init(requestedCount: Int, resultCount: Int) {
    self.requestedCount = requestedCount
    self.resultCount = resultCount
  }
}

public class Box {
  var r0: Int
  var r1: Int
  var g0: Int
  var g1: Int
  var b0: Int
  var b1: Int
  var vol: Int

  init(r0: Int = 0, r1: Int = 0, g0: Int = 0, g1: Int = 0, b0: Int = 0, b1: Int = 0, vol: Int = 0) {
    self.r0 = r0
    self.r1 = r1
    self.g0 = g0
    self.g1 = g1
    self.b0 = b0
    self.b1 = b1
    self.vol = vol
  }

  public var description: String {
    return "Box R \(r0) -> \(r1) G \(g0) -> \(g1) B \(b0) -> \(b1) VOL = \(vol)"
  }
}

File: python/quantize/quantizer_wu.py
# /**
#  * @license
#  * Copyright 2021 Google LLC
#  *
#  * Licensed under the Apache License, Version 2.0 (the "License");
#  * you may not use this file except in compliance with the License.
#  * You may obtain a copy of the License at
#  *
#  *      http://www.apache.org/licenses/LICENSE-2.0
#  *
#  * Unless required by applicable law or agreed to in writing, software
#  * distributed under the License is distributed on an "AS IS" BASIS,
#  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  * See the License for the specific language governing permissions and
#  * limitations under the License.
#  */

from utils.color_utils import *
from quantize.quantizer_map import *

INDEX_BITS = 5
SIDE_LENGTH = 33 # ((1 << INDEX_INDEX_BITS) + 1)
TOTAL_SIZE = 35937 # SIDE_LENGTH * SIDE_LENGTH * SIDE_LENGTH
directions = {
    "RED" : 'red',
    "GREEN" : 'green',
    "BLUE": 'blue',
}

# /**
#  * An image quantizer that divides the image's pixels into clusters by
#  * recursively cutting an RGB cube, based on the weight of pixels in each area
#  * of the cube.
#  *
#  * The algorithm was described by Xiaolin Wu in Graphic Gems II, published in
#  * 1991.
#  */
class QuantizerWu:
    def __init__(self, weights = [], momentsR = [], momentsG = [], momentsB = [], moments = [], cubes = []):
        self.weights = weights
        self.momentsR = momentsR
        self.momentsG = momentsG
        self.momentsB = momentsB
        self.moments = moments
        self.cubes = cubes

    # /**
    #  * @param pixels Colors in ARGB format.
    #  * @param maxColors The number of colors to divide the image into. A lower
    #  *     number of colors may be returned.
    #  * @return Colors in ARGB format.
    #  */
    def quantize(self, pixels, maxColors):
        self.constructHistogram(pixels)
        self.computeMoments()
        createBoxesResult = self.createBoxes(maxColors)
        results = self.createResult(createBoxesResult.resultCount)
        return results

    def constructHistogram(self, pixels):
        _a = None
        self.weights = [0] * TOTAL_SIZE
        self.momentsR = [0] * TOTAL_SIZE
        self.momentsG = [0] * TOTAL_SIZE
        self.momentsB = [0] * TOTAL_SIZE
        self.moments = [0] * TOTAL_SIZE
        countByColor = QuantizerMap.quantize(pixels)
        for (pixel, count) in countByColor.items():
            red = redFromArgb(pixel)
            green = greenFromArgb(pixel)
            blue = blueFromArgb(pixel)
            bitsToRemove = 8 - INDEX_BITS
            iR = (red >> bitsToRemove) + 1
            iG = (green >> bitsToRemove) + 1
            iB = (blue >> bitsToRemove) + 1
            index = self.getIndex(iR, iG, iB)
            self.weights[index] = (self.weights[index] if len(self.weights) > index else 0) + count
            self.momentsR[index] += count * red
            self.momentsG[index] += count * green
            self.momentsB[index] += count * blue
            self.moments[index] += count * (red * red + green * green + blue * blue)

    def computeMoments(self):
        for r in range(1, SIDE_LENGTH):
            area = [0] * SIDE_LENGTH
            areaR = [0] * SIDE_LENGTH
            areaG = [0] * SIDE_LENGTH
            areaB = [0] * SIDE_LENGTH
            area2 = [0.0] * SIDE_LENGTH
            for g in range(1, SIDE_LENGTH):
                line = 0
                lineR = 0
                lineG = 0
                lineB = 0
                line2 = 0.0
                for b in range(1, SIDE_LENGTH):
                    index = self.getIndex(r, g, b)
                    line += self.weights[index]
                    lineR += self.momentsR[index]
                    lineG += self.momentsG[index]
                    lineB += self.momentsB[index]
                    line2 += self.moments[index]
                    area[b] += line
                    areaR[b] += lineR
                    areaG[b] += lineG
                    areaB[b] += lineB
                    area2[b] += line2
                    previousIndex = self.getIndex(r - 1, g, b)
                    self.weights[index] = self.weights[previousIndex] + area[b]
                    self.momentsR[index] = self.momentsR[previousIndex] + areaR[b]
                    self.momentsG[index] = self.momentsG[previousIndex] + areaG[b]
                    self.momentsB[index] = self.momentsB[previousIndex] + areaB[b]
                    self.moments[index] = self.moments[previousIndex] + area2[b]

    def createBoxes(self, maxColors):
        self.cubes = [Box() for x in [0] * maxColors]
        volumeVariance = [0.0] * maxColors
        self.cubes[0].r0 = 0
        self.cubes[0].g0 = 0
        self.cubes[0].b0 = 0
        self.cubes[0].r1 = SIDE_LENGTH - 1
        self.cubes[0].g1 = SIDE_LENGTH - 1
        self.cubes[0].b1 = SIDE_LENGTH - 1
        generatedColorCount = maxColors
        next = 0
        for i in range(1, maxColors):
            if (self.cut(self.cubes[next], self.cubes[i])):
                volumeVariance[next] = self.variance(self.cubes[next]) if self.cubes[next].vol > 1 else 0.0
                volumeVariance[i] = self.variance(self.cubes[i]) if self.cubes[i].vol > 1 else 0.0
            else:
                volumeVariance[next] = 0.0
                i -= 1
            next = 0
            temp = volumeVariance[0]
            for j in range(1, i):
                if (volumeVariance[j] > temp):
                    temp = volumeVariance[j]
                    next = j
            if (temp <= 0.0):
                generatedColorCount = i + 1
                break
        return CreateBoxesResult(maxColors, generatedColorCount)

    def createResult(self, colorCount):
        colors = []
        for i in range(colorCount):
            cube = self.cubes[i]
            weight = self.volume(cube, self.weights)
            if (weight > 0):
                r = round(self.volume(cube, self.momentsR) / weight)
                g = round(self.volume(cube, self.momentsG) / weight)
                b = round(self.volume(cube, self.momentsB) / weight)
                color = (255 << 24) | ((r & 0x0ff) << 16) | ((g & 0x0ff) << 8) | (b & 0x0ff)
                colors.append(color)
        return colors

    def variance(self, cube):
        dr = self.volume(cube, self.momentsR)
        dg = self.volume(cube, self.momentsG)
        db = self.volume(cube, self.momentsB)
        xx = self.moments[self.getIndex(cube.r1, cube.g1, cube.b1)] - self.moments[self.getIndex(cube.r1, cube.g1, cube.b0)] - self.moments[self.getIndex(cube.r1, cube.g0, cube.b1)] + self.moments[self.getIndex(cube.r1, cube.g0, cube.b0)] - self.moments[self.getIndex(cube.r0, cube.g1, cube.b1)] + self.moments[self.getIndex(cube.r0, cube.g1, cube.b0)] + self.moments[self.getIndex(cube.r0, cube.g0, cube.b1)] - self.moments[self.getIndex(cube.r0, cube.g0, cube.b0)]
        hypotenuse = dr * dr + dg * dg + db * db
        volume = self.volume(cube, self.weights)
        return xx - hypotenuse / volume

    def cut(self, one, two):
        wholeR = self.volume(one, self.momentsR)
        wholeG = self.volume(one, self.momentsG)
        wholeB = self.volume(one, self.momentsB)
        wholeW = self.volume(one, self.weights)
        maxRResult = self.maximize(one, directions["RED"], one.r0 + 1, one.r1, wholeR, wholeG, wholeB, wholeW)
        maxGResult = self.maximize(one, directions["GREEN"], one.g0 + 1, one.g1, wholeR, wholeG, wholeB, wholeW)
        maxBResult = self.maximize(one, directions["BLUE"], one.b0 + 1, one.b1, wholeR, wholeG, wholeB, wholeW)
        direction = None
        maxR = maxRResult.maximum
        maxG = maxGResult.maximum
        maxB = maxBResult.maximum
        if (maxR >= maxG and maxR >= maxB):
            if (maxRResult.cutLocation < 0):
                return False
            direction = directions["RED"]
        elif (maxG >= maxR and maxG >= maxB):
            direction = directions["GREEN"]
        else:
            direction = directions["BLUE"]
        two.r1 = one.r1
        two.g1 = one.g1
        two.b1 = one.b1

        if (direction == directions["RED"]):
            one.r1 = maxRResult.cutLocation
            two.r0 = one.r1
            two.g0 = one.g0
            two.b0 = one.b0
        elif (direction == directions["GREEN"]):
            one.g1 = maxGResult.cutLocation
            two.r0 = one.r0
            two.g0 = one.g1
            two.b0 = one.b0
        elif (direction == directions["BLUE"]):
            one.b1 = maxBResult.cutLocation
            two.r0 = one.r0
            two.g0 = one.g0
            two.b0 = one.b1
        else:
            raise Exception('unexpected direction ' + direction)

        one.vol = (one.r1 - one.r0) * (one.g1 - one.g0) * (one.b1 - one.b0)
        two.vol = (two.r1 - two.r0) * (two.g1 - two.g0) * (two.b1 - two.b0)
        return True

    def maximize(self, cube, direction, first, last, wholeR, wholeG, wholeB, wholeW):
        bottomR = self.bottom(cube, direction, self.momentsR)
        bottomG = self.bottom(cube, direction, self.momentsG)
        bottomB = self.bottom(cube, direction, self.momentsB)
        bottomW = self.bottom(cube, direction, self.weights)
        max = 0.0
        cut = -1
        halfR = 0
        halfG = 0
        halfB = 0
        halfW = 0
        for i in range(first, last):
            halfR = bottomR + self.top(cube, direction, i, self.momentsR)
            halfG = bottomG + self.top(cube, direction, i, self.momentsG)
            halfB = bottomB + self.top(cube, direction, i, self.momentsB)
            halfW = bottomW + self.top(cube, direction, i, self.weights)
            if (halfW == 0):
                continue
            tempNumerator = (halfR * halfR + halfG * halfG + halfB * halfB) * 1.0
            tempDenominator = halfW * 1.0
            temp = tempNumerator / tempDenominator
            halfR = wholeR - halfR
            halfG = wholeG - halfG
            halfB = wholeB - halfB
            halfW = wholeW - halfW
            if (halfW == 0):
                continue
            tempNumerator = (halfR * halfR + halfG * halfG + halfB * halfB) * 1.0
            tempDenominator = halfW * 1.0
            temp += tempNumerator / tempDenominator
            if (temp > max):
                max = temp
                cut = i
        return MaximizeResult(cut, max)

    def volume(self, cube, moment):
        return (moment[self.getIndex(cube.r1, cube.g1, cube.b1)] - moment[self.getIndex(cube.r1, cube.g1, cube.b0)] - moment[self.getIndex(cube.r1, cube.g0, cube.b1)] + moment[self.getIndex(cube.r1, cube.g0, cube.b0)] - moment[self.getIndex(cube.r0, cube.g1, cube.b1)] + moment[self.getIndex(cube.r0, cube.g1, cube.b0)] + moment[self.getIndex(cube.r0, cube.g0, cube.b1)] - moment[self.getIndex(cube.r0, cube.g0, cube.b0)])

    def bottom(self, cube, direction, moment):
        if (direction == directions["RED"]):
            return (-moment[self.getIndex(cube.r0, cube.g1, cube.b1)] + moment[self.getIndex(cube.r0, cube.g1, cube.b0)] + moment[self.getIndex(cube.r0, cube.g0, cube.b1)] - moment[self.getIndex(cube.r0, cube.g0, cube.b0)])
        elif (direction == directions["GREEN"]):
            return (-moment[self.getIndex(cube.r1, cube.g0, cube.b1)] + moment[self.getIndex(cube.r1, cube.g0, cube.b0)] + moment[self.getIndex(cube.r0, cube.g0, cube.b1)] - moment[self.getIndex(cube.r0, cube.g0, cube.b0)])
        elif (direction == directions["BLUE"]):
            return (-moment[self.getIndex(cube.r1, cube.g1, cube.b0)] + moment[self.getIndex(cube.r1, cube.g0, cube.b0)] + moment[self.getIndex(cube.r0, cube.g1, cube.b0)] - moment[self.getIndex(cube.r0, cube.g0, cube.b0)])
        else:
            raise Exception('unexpected direction ' + direction)

    def top(self, cube, direction, position, moment):
        if (direction == directions["RED"]):
            return (moment[self.getIndex(position, cube.g1, cube.b1)] - moment[self.getIndex(position, cube.g1, cube.b0)] - moment[self.getIndex(position, cube.g0, cube.b1)] + moment[self.getIndex(position, cube.g0, cube.b0)])
        elif (direction == directions["GREEN"]):
            return (moment[self.getIndex(cube.r1, position, cube.b1)] - moment[self.getIndex(cube.r1, position, cube.b0)] - moment[self.getIndex(cube.r0, position, cube.b1)] + moment[self.getIndex(cube.r0, position, cube.b0)])
        elif (direction == directions["BLUE"]):
            return (moment[self.getIndex(cube.r1, cube.g1, position)] - moment[self.getIndex(cube.r1, cube.g0, position)] - moment[self.getIndex(cube.r0, cube.g1, position)] + moment[self.getIndex(cube.r0, cube.g0, position)])
        else:
            raise Exception('unexpected direction ' + direction)

    def getIndex(self, r, g, b):
        return (r << (INDEX_BITS * 2)) + (r << (INDEX_BITS + 1)) + r + (g << INDEX_BITS) + g + b

# /**
#  * Keeps track of the state of each box created as the Wu  quantization
#  * algorithm progresses through dividing the image's pixels as plotted in RGB.
#  */
class Box:
    def __init__(self, r0 = 0, r1 = 0, g0 = 0, g1 = 0, b0 = 0, b1 = 0, vol = 0):
        self.r0 = r0
        self.r1 = r1
        self.g0 = g0
        self.g1 = g1
        self.b0 = b0
        self.b1 = b1
        self.vol = vol

# /**
#  * Represents final result of Wu algorithm.
#  */
class CreateBoxesResult:
    # /**
    #  * @param requestedCount how many colors the caller asked to be returned from
    #  *     quantization.
    #  * @param resultCount the actual number of colors achieved from quantization.
    #  *     May be lower than the requested count.
    #  */
    def __init__(self, requestedCount, resultCount):
        self.requestedCount = requestedCount
        self.resultCount = resultCount

# /**
#  * Represents the result of calculating where to cut an existing box in such
#  * a way to maximize variance between the two new boxes created by a cut.
#  */
class MaximizeResult:
    def __init__(self, cutLocation, maximum):
        self.cutLocation = cutLocation
        self.maximum = maximum
